WordPress/gutenberg
# Add new package `@wordpress/lazy-import` for lazily installed packages

This pull request seeks to introduce a new package `@wordpress/lazy-import`, which can be used to dynamically install modules on-demand. The proposed usefulness here is in cases where modules are used in ad-hoc scripts (see included changes for benchmarking dependencies) or in cases where the dependency is only used in particular logic paths which may not always be encountered, or at least not frequently enough to warrant the up-front cost of the install. `@wordpress/scripts` is a good candidate for this, where this package is very similar to what was done in #20215, but extrapolated for _any_ dependency.

As mentioned, the changes include removal of top-level dependencies for various shallow-equals implementations which are used only for the (rarely-run) benchmarking script bundled with `@wordpress/is-shallow-equal`.

Future potential usage could include lazily importing individual dependencies of `@wordpress/script` on a per-command basis (e.g. install Jest only if and when running `test-unit`, Prettier when running `format-js`, ESLint when running `lint-js`, etc).

**Implementation Notes:**

The [lazy-require](https://github.com/bevry/lazy-require) package served as some inspiration for this one. However, aside from the fact that there are some issues with it (https://github.com/bevry/lazy-require/issues/48), it does not support an option to specify a version of the package to use. This can be problematic, since it could potentially mean that a later major version update of a package could break expected usage.

With this implementation, the usage is such that you can call it with an argument string as would be passed to `npm install`, including an optional version specifier:

```js
await lazyImport( 'fbjs@^1.0.0' );
```

This is partly inspired by [RunKit](https://runkit.com), which works this way when using `require`:

https://runkit.com/embed/qxtd3i7j9org

Internally, it uses the same modules used by NPM to parse and validate the version specifier, so it should act exactly as one should expect as if the string is passed to `npm install`.

Normally, it is not possible to install multiple versions of the same package to a local project. To work around this limitation, the implementation (ab)uses [npm aliases](https://github.com/npm/rfcs/blob/latest/implemented/0001-package-aliases.md), where a unique name is generated using in the input argument.

For illustration purposes, given an input `shallow-equal@^1.2.1`, it behaves like:

```
npm install --no-save @wordpress/lazy-import.5ab856344c555e50779f56b4cd1f2e02@npm:shallow-equal@^1.2.1
```

...where `5ab856344c555e50779f56b4cd1f2e02` is an md5 hash of the input `shallow-equal@^1.2.1`.

**Testing Instructions:**

Verify that the benchmarking script still runs without issue, even after clearing `node_modules` and installing with the benchmark dependencies removed:

```
node packages/is-shallow-equal/benchmark/index.js
```

The first run may take some additional time since the dependencies must be installed on-demand. Subsequent runs should start faster (though still delayed, since benchmarking itself takes some time to run).
